src/runtime/objmodel.cpp

```diff
@@ -1705,8 +1706,37 @@ void setattrInternal(Box* obj, const std::string& attr, Box* val, SetattrRewrite
         // We don't need to to the invalidation stuff in this case.
         return;
     } else {
+        // Finally, check __setattr__
+        if (obj->cls->tp_setattr) {
+            rewrite_args = NULL;
+            REWRITE_ABORTED("");
+
+            int rtn = obj->cls->tp_setattr(obj, const_cast<char*>(attr.c_str()), val);
+            if (rtn)
+                throwCAPIException();
+            return;
+        }
+        Box* setattr = typeLookup(obj->cls, setattr_str, NULL);
+        if (setattr) {
+            rewrite_args = NULL;
+            REWRITE_ABORTED("");
+
+            // if we're dealing with a BoxedWrapperDescriptor wrapping
+            // PyObject_GenericSetAttr, skip calling __setattr__, as
+            // that will just re-enter us.
+            if (setattr->cls != wrapperdescr_cls
+                || ((BoxedWrapperDescriptor*)setattr)->wrapped != PyObject_GenericSetAttr) {
+                Box* boxstr = boxString(attr);
+                Box* rtn = callattrInternal2(obj, &setattr_str, CLASS_ONLY, NULL, ArgPassSpec(2), boxstr, val);
+                if (rtn == NULL)
+                    runtimeCall3(setattr, ArgPassSpec(3), obj, boxstr, val);
```

Comment:
you're right about the fallback not being necessary.  that check predates the addition of the typeLookup above above.  Going to switch from runtimeCall3 to runtimeCallInternal to match the rest of the code in this method.